Improved message dispatch to pallets

vgeddes opened the pull request:
Solidity apps can now call any dispatchable function on any substrate pallet, allowing for greater flexibility in messaging.

Changes:
* Added `artemis-dispatch` pallet, inspired by the [call-dispatch](https://github.com/paritytech/parity-bridges-common/blob/master/modules/call-dispatch/src/lib.rs) module in in parity-bridges-common
* Added a custom Origin for calls dispatched on behalf of a remote contract. Used by apps for authorization.
* Removed obsolete `Application` trait.
* Updated Solidity ETH and ERC20 contracts to encode Call objects
* Updated Substrate ETH and ERC20 pallets now have `mint` dispatchables.

Other changes which somehow crept in:
* Fixed nonce handling (#241)

Closes: #239 #241

The E2E tests pass successfully with these changes

musnit commented on ethereum/contracts/ETHApp.sol:
```diff
@@ -80,17 +74,11 @@ contract ETHApp {
-                payload.recipient,
-                payload.amount.toBytes32LE()
-            );
+        return abi.encodePacked(CALL_INDEX, _sender, _recipient, _amount.toBytes32LE());
```
do you have a guess for how much this adds to the cost of the transaction in USD?

vgeddes replied:
You mean the new SCALE-encoding vs older ABI-encoding?

Well I ran a benchmark and the SCALE-encoding in this PR adds 6.64% to the gas used (76240 vs 81307) by the `lock()` transaction. At current gas prices (average confirmation times) that adds an extra $1.77 to the cost of the transaction, coming up to a total of $26.64879 according to https://ethgasstation.info/calculatorTxV.php.

vgeddes replied:
Though gas prices are unusually high today, roughly 50% more than the average over the past 4 months.

musnit replied:
you mean 6% just for adding the  CALL_INDEX parameter? or 6% in total compared to abi encoding?

vgeddes replied:
the latter, 6% in total compared to abi encoding.

vgeddes replied:
I've further optimized the SCALE-encoding of `uint256`, so its now about 21.5% faster than the previous algorithm. So total  overhead should now only be about 5% instead of 6%.

vgeddes commented:
@philipstanislaus 

I've improved nonces. Here's how it works now in pseudocode:

Outbound channel:
```
submit(message):
  nonce = nonce + 1
  send_message(message, nonce)
```

Inbound channel:
```
receive(message):
   if message.nonce != nonce + 1:
      raise Error
   nonce = nonce + 1
```

So consumed nonce's now start at 1 instead of 0, and the outbound channel tracks the last consumed nonce, and the inbound channel tracks the nonce for the last received message. Hope this addresses your concerns?

musnit reviewed (approved):
looks good overall - lets merge to unblock other work and can address comments from other reviewers in follow up PR
